Classification/FeatureSelection/src/main/java/org/tribuo/classification/fs/FSMatrix.java

```diff
@@ -0,0 +1,191 @@
+
+    /**
+     * Calculates the joint mutual information I(X_iX_j;X_k)
+     * @param firstIndex The index i
+     * @param jointIndex The index j
+     * @param targetIndex The index k
+     * @return I(X_iX_j;X_k)
+     */
+    double jmi(int firstIndex, int jointIndex, int targetIndex);
+
+    /**
+     * Generates an array of I(X_iZ;Y), ranging X_i over all possible features, where Y is the label.
+     * @param jointIndex The index of the Z variable
+     * @return An array where array[i] = I(X_iZ;Y)
+     */
+    default double[] jmiList(int jointIndex) {
+        int numFeatures = getNumFeatures();
+        double[] row = new double[numFeatures];
+        for (int i = 0; i < numFeatures; i++) {
+            row[i] = jmi(i,jointIndex);
+        }
+        return row;
+    }
+
+    /**
+     * Generates an array of I(X_iZ;Y), ranging X_i over all possible features.
+     * @param targetIndex The index of the Y variable
+     * @param jointIndex The index of the Z variable
+     * @return An array where array[i] = I(X_iZ;Y)
+     */
+    default double[] jmiList(int targetIndex, int jointIndex) {
+        int numFeatures = getNumFeatures();
+        double[] row = new double[numFeatures];
+        for (int i = 0; i < numFeatures; i++) {
+            row[i] = jmi(i,jointIndex,targetIndex);
+        }
+        return row;
+    }
+
+    /**
+     * Calculates the conditional mutual information I(X_i;Y|X_j), where Y is the label
+     * @param featureIndex The index i
+     * @param conditionIndex The index j
+     * @return I(X_i;Y|X_j)
+     */
+    double cmi(int featureIndex, int conditionIndex);
+
+    /**
+     * Calculates the conditional mutual information I(X_i;X_j|X_k)
+     * @param firstIndex The index i
+     * @param secondIndex The index j
+     * @param conditionIndex The index k
+     * @return I(X_i;X_j|X_k)
+     */
+    double cmi(int firstIndex, int secondIndex, int conditionIndex);
+
+    /**
+     * Generates an array of I(X_i;Y|Z), ranging X_i over all possible features, where Y is the label.
+     * @param conditionIndex The index of the Z variable
+     * @return An array where array[i] = I(X_i;Y|Z)
+     */
+    default double[] cmiList(int conditionIndex) {
+        int numFeatures = getNumFeatures();
+        double[] row = new double[numFeatures];
+        for (int i = 0; i < numFeatures; i++) {
+            row[i] = cmi(i,conditionIndex);
+        }
+        return row;
+    }
+
+    /**
+     * Generates an array of I(X_i;Y|Z), ranging X_i over all possible features.
+     * @param targetIndex The index of the Y variable
+     * @param conditionIndex The index of the Z variable
+     * @return An array where array[i] = I(X_i;Y|Z)
+     */
+    default double[] cmiList(int targetIndex, int conditionIndex) {
+        int numFeatures = getNumFeatures();
+        double[] row = new double[numFeatures];
+        for (int i = 0; i < numFeatures; i++) {
+            row[i] = cmi(i,targetIndex,conditionIndex);
```

Comment:
Inconsistent naming between this and the singleton version
```suggestion
     * Generates an array of I(X_i;X_j|X_k), ranging X_i over all possible features.
     * @param featureIndex The index of the X_j variable
     * @param conditionIndex The index of the X_k variable
     * @return An array where array[i] = I(X_i;X_j|X_k)
     */
    default double[] cmiList(int featureIndex, int conditionIndex) {
        int numFeatures = getNumFeatures();
        double[] row = new double[numFeatures];
        for (int i = 0; i < numFeatures; i++) {
            row[i] = cmi(i,featureIndex,conditionIndex);
```